styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesPodEventTranslator.java

```diff
@@ -131,101 +131,118 @@ public TerminationLogMessage(
-          generatedEvents.add(hasError.get());
-          break;
-
-        default:
-          // no event
-          break;
-      }
+    final Optional<ContainerStatus> mainContainerStatusOpt = getMainContainerStatus(pod);
 
+    final Optional<Event> hasError = isInErrorState(workflowInstance, pod, mainContainerStatusOpt);
+    if (hasError.isPresent()) {
+      handleError(state, generatedEvents, hasError.get());
       return generatedEvents;
     }
 
-    final PodStatus status = pod.getStatus();
-    final String phase = status.getPhase();
+    if (isStarted(pod, mainContainerStatusOpt)) {
+      handleStarted(workflowInstance, state, generatedEvents);
+    }
 
-    boolean exited = false;
-    boolean started = false;
+    if (isExited(pod, mainContainerStatusOpt)) {
+      handleExited(workflowInstance, state, pod, mainContainerStatusOpt, stats, generatedEvents);
+    }
 
-    final Optional<ContainerStatus> containerStatus = getMainContainerStatus(pod);
-    switch (phase) {
+    return generatedEvents;
+  }
+
+  private static void handleExited(WorkflowInstance workflowInstance, RunState state,
+                                   Pod pod, Optional<ContainerStatus> mainContainerStatusOpt,
+                                   Stats stats,
+                                   List<Event> generatedEvents) {
+    switch (state.state()) {
+      case PREPARE:
+      case SUBMITTED:
+        generatedEvents.add(Event.started(workflowInstance));
+        // intentional fall-through
+
+      case RUNNING:
+        final Optional<Integer> exitCode = mainContainerStatusOpt.flatMap(cs ->
+            getExitCodeIfValid(workflowInstance.toKey(), pod, cs, stats));
+        generatedEvents.add(Event.terminate(workflowInstance, exitCode));
+        break;
+
+      default:
+        // no event
+        break;
+    }
+  }
+
+  private static void handleStarted(final WorkflowInstance workflowInstance, final RunState state,
+                                    final List<Event> generatedEvents) {
+    switch (state.state()) {
+      case PREPARE:
+      case SUBMITTED:
+        generatedEvents.add(Event.started(workflowInstance));
+        break;
+
+      default:
+        // no event
+        break;
+    }
+  }
+
+  private static void handleError(RunState state, List<Event> generatedEvents, Event event) {
+    switch (state.state()) {
+      case PREPARE:
+      case SUBMITTED:
+      case RUNNING:
+        generatedEvents.add(event);
+        break;
+
+      default:
+        // no event
+        break;
+    }
+  }
+
+  private static boolean isExited(Pod pod, Optional<ContainerStatus> mainContainerStatusOpt) {
+    switch (pod.getStatus().getPhase()) {
       case "Running":
         // Check if the main container has exited
-        exited = containerStatus.map(ContainerStatus::getState)
-                                .map(ContainerState::getTerminated)
-                                .isPresent();
-        if (exited) {
-          break;
```

Comment:
This break had the effect that if `exited == true` then `started == false` regardless of the container status. Is that invariant ok to remove?

Reply:
I think they are mutually exclusive. Will double check.